File: algorithm/python/88_merge_sorted_array.py
# Solution 1

class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """
        if n == 0:
            pass
        elif m == 0:
            # Remove extra zeros at the end
            # 去除末尾多余的0
            for tmp in range(n):
                nums1.pop()
            for i in range(n):
                nums1.append(nums2[i])
        else:
            for tmp in range(n):
                nums1.pop()
            i = j = 0
            while i != m and j != n:
                if nums1[i] >= nums2[j]:
                    nums1.insert(i, nums2[j])
                    i += 1
                    j += 1
                    # a new element is inserted, the length of nums1 is increased by one, that is, m = m + 1
                    # 因为插入了新元素，nums1长度要增加一，即m=m+1
                    m += 1 
                else:
                    i += 1
            if i == m:
                for k in range(j, n):
                    nums1.append(nums2[k])


# Solution 2, easy to understand, but poor efficiency

class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """
        if n == 0:
            pass
        else:
            # Remove extra zeros at the end
            # 去除末尾多余的0
            for tmp in range(n):
                nums1.pop()
            for i in range(n):
                nums1.append(nums2[i])
            nums1.sort()

'''
    This is my personal record of solving Leetcode Problems. 
    If you have any questions, please discuss them in [Issues](https://github.com/mengxinayan/leetcode/issues).
    Copyright (C) 2020  mengxinayan

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''

File: algorithm/java/88_merge_sorted_array.java
// Solution 1. allocate a new array to store sorted result.

class Solution {
    public void merge(int[] nums1, int m, int[] nums2, int n) {
        int[] sortedArr = new int[m+n];
        int i = 0;
        int j = 0;
        while (i < m && j < n) {
            if (nums1[i] <= nums2[j]) {
                sortedArr[i+j] = nums1[i];
                i++;
            } else {
                sortedArr[i+j] = nums2[j];
                j++;
            }
        }
        for (; i < m; i++) {
            sortedArr[i+j] = nums1[i];
        }
        for (; j < n; j++) {
            sortedArr[i+j] = nums2[j];
        }
        for (int k = 0; k < m+n; k++) {
            nums1[k] = sortedArr[k];
        }
        return ;
    }
}


// Solution 2 insert number at the end of nums1 

class Solution {
    public void merge(int[] nums1, int m, int[] nums2, int n) {
        int i = m-1;
        int j = n-1;
        int k = m + n - 1;
        while (i >= 0 && j >= 0) {
            if (nums1[i] >= nums2[j]) {
                nums1[k] = nums1[i];
                i--;
            } else {
                nums1[k] = nums2[j];
                j--;
            }
            k--;
        }
        for (; i >= 0; i--, k--) {
            nums1[k] = nums1[i];
        }
        for (; j >= 0; j--, k--) {
            nums1[k] = nums2[j];
        }
        return ;
    }
}

/**
 * This is my personal record of solving Leetcode Problems. 
 * If you have any questions, please discuss them in [Issues](https://github.com/mengxinayan/leetcode/issues).
 * Copyright (C) 2020-2022  mengxinayan
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
